contractsV5/FeeHandler.sol

```diff
@@ -0,0 +1,189 @@
+
+    // Todo: future optimisation to accumulate rebates for 2 rebate wallet
+    // encode totals, 128 bits per reward / rebate
+    function handleFees(address[] calldata eligibleWallets, uint[] calldata rebatePercentages) external payable onlyKyberNetwork returns(bool) {
+        // Decoding BRR data
+        (uint burnInBPS, uint rewardInBPS, uint epoch, uint expiryBlock) = decodeData();
+        uint rebateInBPS = BPS - rewardInBPS - burnInBPS;
+
+        // Check current block number
+        if(block.number > expiryBlock) {
+            (burnInBPS, rewardInBPS, rebateInBPS, epoch, expiryBlock) = kyberDAOContract.getLatestBRRData();
+
+            // Update brrAndEpochData
+            brrAndEpochData = encodeData(burnInBPS, rewardInBPS, epoch, expiryBlock);
+        }
+
+        uint rebateWei = rebateInBPS * msg.value / BPS;
+        uint rewardWei = rewardInBPS * msg.value / BPS;
+        for(uint i = 0; i < eligibleWallets.length; i ++) {
+            // Internal accounting for rebates per reserve wallet (totalRebatesPerRebateWallet)
+            totalRebatesPerRebateWallet[eligibleWallets[i]] = rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for total rebates (totalRebates)
+            totalRebates += rebateWei * rebatePercentages[i] / 100;
+            // Internal accounting for reward per epoch (totalRewardsPerEpoch)
+            totalRewardsPerEpoch[epoch] += rewardWei;
+            // Internal accounting for total rewards (totalRewards)
+            totalRewards += rewardWei;
+
+            emit AccumulateReserveRebate(eligibleWallets[i], rebateWei * rebatePercentages[i] / 100);
+        }
+        return true;
+    }
+
+    event DistributeRewards(address staker, uint amountWei);
+
+    function claimStakerReward(address staker, uint percentageInPrecision, uint epoch) public onlyDAO returns(uint) {
+        // Amount of reward to be sent to staker
+        uint amount = totalRewardsPerEpoch[epoch] * percentageInPrecision / 10**18;
+
+        // Update total rewards and total rewards per epoch
+        totalRewardsPerEpoch[epoch] -= amount;
+        totalRewards -= amount;
+
+        // send reward to staker
+        (bool success, ) = staker.call.value(amount)("");
+        require(success, "Transfer of rewards to staker failed.");
+
+        emit DistributeRewards(staker, amount);
+
+        return amount;
+
+    }
+
+    event DistributeRebate(address rebateWallet, uint amountWei);
+
+    // Using rebateWallet instead of reserve so I don't have to store KyberNetwork variable.
+    function claimReserveRebate(address rebateWallet) public returns (uint){
+        // Get total amount of rebate accumulated
+        uint amount = totalRebatesPerRebateWallet[rebateWallet] - 1;
+
+        // Update total rebate and rebate per rebate wallet amounts
+        totalRebates -= amount;
+        totalRebatesPerRebateWallet[rebateWallet] = 1; // Do not use 0 to avoid potential issues with div by 0
```

Comment:
I was thinking not to use 0 to avoid paying high gas when setting back from 0 to non zero.

Reply:
noted, updated and will push fix when other comments are fixed